### PR TITLE
[cae760] Frame has non-empty accessible name

### LINKED ISSUE #1170: What does it actually mean to be part of sequential focus navigation? (cae760)
Running through the test cases changed in #1102, I took note of failed example 7:

```html
<iframe title=" " src="/test-assets/SC4-1-2-frame-doc.html" role="none" tabindex="0"></iframe>
```

This is supposed to fail the rule and this I'm absolutely on board with; the contents of the `<iframe>` can still be tabbed to and so one would need to name the `<iframe>`. However, [Alfa](https://github.com/siteimprove/alfa) doesn't fail this case because it interprets "part of sequential focus navigation" to mean "elements returned from the [sequential navigation search algorithm](https://html.spec.whatwg.org/multipage/interaction.html#sequential-navigation-search-algorithm)". This algorithm is used for determining what should happen when a user presses the tab button on their keyboard; given their current position and a direction, what should the next active focusable area be? Per this algorithm, browsing context containers, which `<iframe>` elements are, redirect focus to their contents, and are therefore never themselves returned from this algorithm. To see the effect of this, try tabbing to the button named "First" in this document and then hit tab:

```html
<button>First</button>
<iframe srcdoc="<button>Second</button>"></iframe>
```

You'll notice that the next active element becomes the button named "Second" and not the `<iframe>` element itself. This holds true even if we attempt to add something like `tabindex="0"` to the `<iframe>` element; the sequential navigation search algorithm doesn't care.

However, `<iframe>` elements _are_ [focusable areas](https://html.spec.whatwg.org/multipage/interaction.html#focusable-area) and are actually even listed of examples as such. This can be verified by doing this:

```js
const iframe = document.querySelector("iframe");
iframe.focus();
console.log(document.activeElement);
```

If you run that code on the document above, you'll also notice that upon pressing tab, the button named "Second" will indeed become the next active element.

So, while a user cannot ever tab directly to an `<iframe>` element, we can programmatically move focus to it. As such, Alfa does not consider it part of sequential focus navigation. However, one could argue that an `<iframe>` element is still part of sequential focus navigation as it _is_ a focusable area that unless it has a negative tab index will be in the [sequential focus navigation _order_](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation-order).

What are peoples thoughts on this? At the very least, there's clearly an ambiguity that needs to be resolved.

### LINKED ISSUE #1367: iframe has non-empty name - problems with examples and applicability [cae760]
Each browser / AT combination presents a slightly different user interface for navigating iframes:

- VoiceOver/Safari 13 - exposes iframe as container in normal reading order named by title, but iframe contents not in until user presses Control-Option-Shift-Down Arrow to enter iframe, first item inside iframe focused by tab key 
- JAWS/IE11 - includes iframe contents in normal reading order, first item inside iframe focused by tab key and frame title not read
- JAWS/FF68 - includes iframe contents in normal reading order, first item inside iframe focused by tab key and frame title is read
- JAWS/Chrome - includes iframe contents in normal reading order, first item inside iframe focused by tab key and frame title is read
- NVDA/IE11 - includes iframe contents in normal reading order, first item inside iframe focused by tab key
- NVDA/FF68 - includes iframe contents in normal reading order, iframe is focusable but always silent (not named by title/acc-name)
- NVDA/Chrome - unable to test due to focus trap in browser address bar

On Windows iframe title is read when tabbing, but not in normal screen reader reading order
On Mac iframe title is read in normal screen reader reading order, but not when tabbing

Test results for the examples are

Example        | VO/Safari 13 | JAWS/IE11 | JAWS/FF68 | JAWS/Chrome | NVDA/FF68 | NVDA/Chrome | NVDA/IE11
---------------|--------------|-----------|-----------|-------------|-----------|-------------|----------
Passed 1       | OK           | OK        | OK        | OK          | **Fail**  | Untested    | OK
Passed 2       | **Fail**     | OK        | OK        | OK          | **Fail**  | Untested    | OK
Passed 3       | **Fail**     | OK        | OK        | OK          | **Fail**  | Untested    | OK
Failed 1       | OK           | OK        | Fail      | OK          | Fail      | Untested    | OK
Failed 2       | Fail         | OK        | Fail      | OK          | Fail      | Untested    | OK
Failed 3       | Fail         | OK        | Fail      | OK          | Fail      | Untested    | OK
Failed 4       | Fail         | OK        | Fail      | OK          | Fail      | Untested    | OK
Inapplicable 1 | OK           | OK        | OK        | OK          | OK        | Untested    | OK
Inapplicable 2 | OK           | OK        | OK        | OK          | OK        | Untested    | OK
Inapplicable 3 | **Fail**     | OK        | OK        | **Fail**    | OK     | Untested    | OK

On Safari untitled iframes are hard to use because there's no indication of frame contents, and users often assume they contain adverts. This is bad if the iframe contains something important - I've seen a Safari user unable to find a login form in an iframe for exactly this reason.

Inapplicable Example 3 fails in VoiceOver with Safari
The Safari issue on Inapplicable Example 3 is a consequence of the UI design (frames exposed as nested containers) but excluded from the applicability. The problem happens in normal reading order, not when tabbing between focusable items. I think the applicability needs extended to cover problems with VoiceOver/Safari by removing the following:
>and that can be accessed by sequential focus navigation

https://act-rules.github.io/rules/cae760#applicability

Passed examples 1,2,3 always fail on NVDA/FF68 because tabbing to the frame always produces silence - even if a title or ARIA label is provided. This looks like a bug so I don't think the examples or applicability should change for this.

Passed examples 2,3 always fail on VoiceOver with Safari because it doesn't use acc-name calculation for iframes (but does use the `iframe name` which is not supposed to be human readable). Again this looks like a bug so I don't think the examples of applicability should change for this.

NVDA with Chrome is untested because there's a focus trap in address bar so you can't tab to the page to test.

**tl/dr;** remove "and that can be accessed by sequential focus navigation" from applicability and make Inapplicable Example 3 a fail.

### PR DESCRIPTION
**Pull request by tombrunet**
Adjustments to the rule based on conversation of task force conversations (Feb 16, 2023).

Fixes #1367
Fixes #1170

Note regarding the two above issues. There is no perfect answer here since the browser behaviors are all different. However, the approach in this PR was to ignore cases where it's clear that the author attempts to remove the iframe from the sequential focus navigation.

Need for Call for Review:
This will require a 2 weeks Call for Review << new rule, or substantial changes affecting a large number of test cases, if in doubt, use this. >>

---

### **When creating PR:**

- [x] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [x] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [x] Add yourself (and co-authors) as "Assignees" for PR.
- [x] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [x] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [x] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [x] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

### CODE REVIEW

**Review by carlosapaduarte — APPROVED** [review]

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]:
```suggestion
  - Some browsers instantly redirect focus from `iframe` elements  to the first focusable element inside that iframe. This redirect makes it appear as though the `iframe` never receives focus. This occurs even if the `iframe` has a non-negative `tabindex` [attribute value][].
```

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: I don't think we should mention this

```suggestion
  - Not all browsers redirect focus on `iframe` elements. This ensures that the contents of `iframe` elements can be scrolled and accessed by using the keyboard. This must not be circumvented by using a negative tabindex, as this will make the `iframe` completely inaccessible for keyboard navigation.
```

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: I would suggest this:

```suggestion
- the `iframe` does not have a negative `tabindex` [attribute value][]; and
- the `iframe` does not have an [explicit semantic role][] of `presentation` or `none`.
```

Restated to follow the conditional list writing style, and then using explicit semantic role, so we don't require an exact attribute value match, to account for whitespace and fallback roles.

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: Would you mind moving this note to the background? That feels like a better place for it.

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]:
```suggestion
Due to these inconsistencies, the applicability of this rule is scoped for `iframe` elements for which there is no indication of an attempt to remove them from accessibility consideration.  Whether `iframe` elements that are inapplicable to this rule still require an accessible name varies between browsers.
```

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: This feels like a slightly better test of this. This could be passed because the title attribute is used.

```suggestion
<iframe src="/test-assets/SC4-1-2-frame-doc.html" role="none"> </iframe>
```

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (WilcoFiers) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: Hmm, this one needs a height so that it isn't scrollable. 

https://www.w3.org/WAI/content-assets/wcag-act-rules/testcases/cae760/803b52f32a5d2193683f152aa5e00174d5f8416a.html

**Review by WilcoFiers — DISMISSED** [review]

**Review by Jym77 — CHANGES_REQUESTED** [review]
Agreeing with Wilco's comment.
It also seems that you need to add yourself as an contributor, iirc this is done by modifying [package.json](https://github.com/act-rules/act-rules.github.io/blob/develop/package.json). I suggest you do that bit in a separate PR (can be merged with one Approval), since it shouldn't be blocked by this PR (in case it takes longer than expected).

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (tombrunet) [reply _rules/iframe-non-empty-accessible-name-cae760.md]: Done

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (tombrunet) [reply _rules/iframe-non-empty-accessible-name-cae760.md]: Added

**Comment by tombrunet** [comment]
@Jym77 https://github.com/act-rules/act-rules.github.io/pull/2048 for adding contributor

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (Jym77) [inline _rules/iframe-non-empty-accessible-name-cae760.md]:
```suggestion
The `frame` element is deprecated, this rule does not consider `frame` or `frameset` elements.
```
No need for the "note" marker in the Background section.

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (Jym77) [inline _rules/iframe-non-empty-accessible-name-cae760.md]: **Suggestion:** I feel this is more of a Background note than an Accessibility Support one 🤔 
Maybe phrasing it as:
```suggestion
Due to inconsistencies in handling focus on `iframe`, this rule ignores `iframe` elements for which there is an attempt to hide them from assistive technologies.  Whether `iframe` elements that are inapplicable to this rule still require an accessible name varies between browsers.
```
and move it to the Background section.
(also trying to simplify the phrasing and avoid negations)

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (Jym77) [inline _rules/iframe-non-empty-accessible-name-cae760.md]:
```suggestion
This `iframe` element has an [explicit semantic role][] of `none`.
```
matching the newest Applicability.

**Review by Jym77 — DISMISSED** [review]
Looks pretty good. A few polishing comments.

File: _rules/iframe-non-empty-accessible-name-cae760.md
Comment (tombrunet) [reply _rules/iframe-non-empty-accessible-name-cae760.md]: @WilcoFiers I like the rewording, but wanted to get your thoughts on Accessibility Support section vs Background before accepting this.

Comment (WilcoFiers) [reply _rules/iframe-non-empty-accessible-name-cae760.md]: LGTM 👍

**Comment by tombrunet** [comment]
@Jym77 I think I have everything you suggested resolved, and I re-requested review, but it still shows a change requested, not sure why. Can you re-review? Thanks.

**Comment by Jym77** [comment]
> @Jym77 I think I have everything you suggested resolved, and I re-requested review, but it still shows a change requested, not sure why. Can you re-review? Thanks.

Yes, the review needs to be dimissed…  (Bottom box, "change requested" dropdown, dots menu on the review to dismiss). I've done it.
dismiss review

**Review by WilcoFiers — APPROVED** [review]

**Review by HelenBurge — APPROVED** [review]

**Review by kengdoj — APPROVED** [review]

**Review by Jym77 — APPROVED** [review]

**Comment by Jym77** [comment]
@tombrunet This looks like this is good to go in Call for Review 🎉 
Do you want me to do it?

**Comment by tombrunet** [comment]
@Jym77 I was out on vacation, but it's sent now.

**Review by bruce-usab — APPROVED** [review]

**Review by carlosapaduarte — APPROVED** [review]

**Pull request merged**
